### DISCUSSION TITLE
Controlled, Uncontrolled, DefaultValues, Numerical, Null

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7621] [Q&A] [answered]
A lot to unpack here.
Yes, I go on and on with this controlled/uncontrolled questions. Please excuse me.. I don't know why this is so hard for me to grasp.

**Let's settle one known fact first:**
- The only difference between controlled/uncontrolled is that uncontrolled does not get the `value` prop from RHF.

**Whan I want:**
1. I want to store numbers as numbers, i.e. not the strings I get from TextInput
2. I don't want to initialize my numerical fields with 0
3. Controlled fields cannot be initialized with null or undefined (due to React limitations)
4. Uncontrolled might be able to handle null values better, but I don't know.
5. Sometimes I want to format the presented value (probably best to do onBlur)

**Q1.** How should I initialize a numerical controlled field to empty string if I cannot use `null` or `undefined`?

**Q2.** Should I use uncontrolled fields instead?

**Q3.** How does uncontrolled fields update the value in the TextInput? I suspect that ref is used?
Can I intercept this value-setting process in a custom component?

**If I go with uncontrolled fields:**
Is it possible to pre-register fields even if they are not rendered? I need a place to store field values even for unmounted fields.
If yes, how do I mount such a field at a later stage when it should be shown in the UI? 
**Q4.** should I `register()` the same fieldName again, but this time connect ref and onChange to the mounted component?
**Q5.** or, can I keep the ref from the first register-call and then use this ref when I actually mount the component?

**Q6** I have some proprietary inputFields, which values are complex objects. How would I initialize such fields with these complex objects? I mean, how will my component get hold of the value when it does not arrive as `value`?

**Comment by bluebill1049** [comment]
> Q1. How should I initialize a numerical controlled field to empty string if I cannot use null or undefined?

input default value should always be empty string at latest for native input.

> Q2. Should I use uncontrolled fields instead?

no, you can use either, but I suggest you pick one and stick to it depending on your app.

> Q3. How does uncontrolled fields update the value in the TextInput? I suspect that ref is used?

Controller/useController through component level state update, native input with ref update.

> Is it possible to pre-register fields even if they are not rendered? I need a place to store field values even for unmounted fields.
If yes, how do I mount such a field at a later stage when it should be shown in the UI?

register('whatever')

> Q4. should I register() the same fieldName again, but this time connect ref and onChange to the mounted component?

recommend not to do that.

> Q5. or, can I keep the ref from the first register-call and then use this ref when I actually mount the component?

don't know your usage, sounds over complicated.

> Q6 I have some proprietary inputFields, which values are complex objects. How would I initialize such fields with these complex objects? I mean, how will my component get hold of the value when it does not arrive as value?

use controlled component, or hidden input. The same why you register an input but with dot syntax to represent array or object.

**Comment by bluebill1049** [comment]
- `Controlled component`: react native, a complex component which state is better to isolated within eg data transform, manipulate and etc
- `uncontrolled component`: where everything is built on and web strength, don't forget hidden input, it can go pretty far.
- `defaultValues`: what's populate the form and measure again with the dirty flag
- `Numerical`: I hate to say this but people making their life hard with transform and parse... keep it string keep it simple, do your transform at the submitting level
- `Null`: similar above

**Comment by Moshyfawn** [comment, marked as answer]
I know @bluebill1049's opinion on value types and very much respect it, so I'd suggest keeping them in mind as the RHF solution is based on his views, designed around his experience of working with forms. So, here's just my 2 cents on the topic. I'm in no way trying to go against the maintainer

> Q1. How should I initialize a numerical controlled field to empty string if I cannot use null or undefined?

You don't need to init **controlled** fields with string values as it might over-complicate the form. It was mentioned in a recent controller-themed issue [raised as a valid point](https://github.com/react-hook-form/react-hook-form/issues/7590#issuecomment-1016951596). For **uncontrolled** values it's a different topic, as it comes down to how React / Web in general works. In this case, I'd stick to the value types rules set by those two techs. 

> Q2. Should I use uncontrolled fields instead?

I'd say, choosing uncontrolled inputs over controlled is valid on the Web, not so much in RN due to the way other platforms work. Uncontrolled is easier and more performant on the Web for sure. Idk why React chooses to promote controlled over uncontrolled for Web React inputs handling, honestly.

> Q3. How does uncontrolled fields update the value in the TextInput? I suspect that ref is used?
> Can I intercept this value-setting process in a custom component?

Whenever the question "can I intercept the value-setting" comes up during creating a solution, I most of the time go with controlled API, as it quite literally allows me to control the value as I need. I understand that we're all programmers and it's tech, so there supposed to be no artistic decisions, but for me it's quite literally: "does this case feel like I need more control over it or I can do with handing the handling to the web?" _(there supposed to be a bunch of disclaimers about how React inputs aren't actually the Web inputs etc, but I hope y'all understand where I'm coming from)_

> Q4. should I register() the same fieldName again, but this time connect ref and onChange to the mounted component?
> Q5. or, can I keep the ref from the first register-call and then use this ref when I actually mount the component?

It feels like you're overthinking something but we don't completely get your use-case and that's where the confusion is coming from. If you do uncontrolled component with `register`, there's no need to "pre-register" it. What's the need to do the initial register before the component has been mounted?

> Q6 I have some proprietary inputFields, which values are complex objects. How would I initialize such fields with these complex objects? I mean, how will my component get hold of the value when it does not arrive as value?

Sounds like a controlled component initialized with the top-level object name, where you pass your complex object state bits into the controller `onChange`. See [this CSB](https://codesandbox.io/s/angry-yalow-66sy7?file=/src/Demo.tsx) with a data range picker object handled _(so, a tuple with `Data` values)_. Not exactly a nested object, but the "complex"  value handling concept is the same. Might give you an idea of what I'm talking about.
Please, feel free to ask for some other example as well.

**Reply by bluebill1049** [reply]
Much better and more comprehensive than mine 👍

**Comment by msageryd** [comment]
Wow, thank you so much guys, you rock!

#### There are some important bits in this, which I wasn't aware of. These are great insights:
- uncontrolled fits best with native web components
- do the value type conversion at submit instead of unchange/blur

#### But I'm not sure how to interprete your stance on defaultValues:

@Moshyfawn "You don't need to init controlled fields with string values".
@Bill "input default value should always be empty string at least for native input"

My own test concluded that using null to initialize controlled fields works on iOS and Android devices, but on web the TextField will not clear upon reset.

Try this snack and run it on "web" and "ios". The field is not cleared when it runs on web.
https://snack.expo.dev/@msageryd/forlorn-mixed-nuts

#### About my urge to register fields ahead of rendering

I might be able to rethink this, but to me the simplest solution is to regster all fields from start, and then render them when a UI is needed.

My form definition can include "conditions". A condition consists of rules `when` and outcome `then`. Here is an example.

``` 
conditions: {
    roomIsAbc: {
      when: 'room',
      is: 'Abc',
      then: {
        fields: {
          caveatNotes: {colorBar: 'red'},
          anotherField: {
            set: "new value for anotherField", 
            icon: "ios-eye-off"},
        },
      },
    },
```

If the value of field `room` is set to "Abc", both `caveatNotes` and `anotherField` will be affected. The affected fields might not even be mounted yet, as they might belong to subforms not yet visited.

For each condition (only one condition in the example) I'm mounting a component `<ConditionWatcher/>` which keeps track of each field involved in the `when` clause via `useWatch` and sets the outcome in a separate datastructure which all rendered fields can access via context (for example to show correct icon), this works great.

#### Alternative solution
I could turn this around and not pre-register the fields. 
- The fields that are not mounted from start would have to initialize themselves from my separate data structure upon mounting. This would be an ok solution.

- But.. What if the updated field (i.e. `anotherField`) is not mounted at all before submit? I'd have to augment the submitted values with `{anotherField: "new value for anotherField"}` since anotherField will not exist in the submitted values.

Pre-registering all fields makes things much easier to me, as I don't have to do the above.

#### Question

Will I get into trouble when I "pre-register the fields"? This is conceptually what I'm doing (controlled fields)

**When the form is started:**
```
useForm()
regsiterAllFields  //(register('anotherField'); register('room'); etc
```

**When fields are rendered in the UI:**
```
{field} = useController();

//"name" is already "pre-registered". Will reusing it get me into trouble?
<TextInput name={name} value={field.value} onChange={field.onChange}/>
```

**Comment by Moshyfawn** [comment]
> My own test concluded that using null to initialize controlled fields works on iOS and Android devices, but on web the TextField will not clear upon reset.

You brought up a very good point, actually. And I haven't delved into how that's came to be. To, it seems like the only way to accommodate for the Web platform is to not initialize text field inputs `null` or `undefined`.

Now, the whole thing about validation of different input value types.. My approach is to always transform the value at the validation schema / custom `validate` callback to validate it and format the data `onSubmit`. So, I'd
1. Use, say, [Zod's `.refine()` method](https://github.com/colinhacks/zod#refine) to validate the input's value in the "proper" data type / format for the check.
```js
const schema = z.object({
  numberValue: z.string().refine(value => +value < 10, {
    message: "numberValue should be less than 10"
  })
})
```
Same for the custom `validate` callback:
```js
validate: (value) => {
  if (+value < 10) {
    return "numberValue should be less than 10"
  }
  return true
}
```
2. Format / Transform the submitted values `onSubmit` _(i.e. `date.toISOString()` for Date as string, +numberValue for string as number etc.)_

_Sorry for the late replies, I need to dedicate some free time  to commit to properly answering your "more complex" questions; not always able to immediately find that time. Not trying to ignore you or anything_

**Reply by msageryd** [reply]
Well, that's a new insight. I didn't know I could convert to number by adding a plus-sign. Neat!
Only drawback is that I need to specifically handle empty strings, since `+'' === 0`, which might differ from the intention when a field is left blank instead of explicitly entering `0`.

### About (1)
This is exactly what I'm doing right now in my spike. Converting dataTypes in three separate places; validation, contition-check and submit.

I'm not using Zod or alike. Those libraries are great for defining rules via code by chaining their JS methods. I'm not sure, but they don't seem to be that easy to work with when the rules are completely declarative, i.e. pure JSON.

### About (2).
That's exactly what I want to do. But before I can format/transform a dataType I need to know what dataType I should transform into. This information sits in my `fieldDefinition`, which is an object keyed with fieldNames. These fieldNames has the dot-notation-style names when the data should be saved deeper into the resulting object.

Wouldn't this be a common scenario? I.e. the need to traverse the resulting data object upon submit, and transforming the data field by field based on some kind of external field definition?

If there is no API for this traversal inluding the building of dot-notation fieldNames, I'm actually thinking of enforcing flat structures in my fieldDefinitions. If I'd handle traversal and name-building myself there is a risk that I'm doing it differently than RHF does internally. I'm worried I'd end up with another format for the dot-notation in some edge cases.

### About *late replies*. 
Both you and Bill has answered all of my questions more swiftly that anyone could expect. I'm really greatful for this.


@bluebill1049 If there is no API for traversing avd transforming submit-data, I think the library could really use such an API. In my scenario I would use such an API like this (`transform` might come from useForm, or be a standalone utility function):

The function would be a recursive tranersing function. onTransform would only be triggered on registered fields, this would prohibit traversal into deep proprietary objects such as custom multiselects and alike.

```
  const onTransform = ({name, value}) => {
    const dataType = fieldDefinitions[name].dataType;
    return dataTypes[dataType].toNativeDataType(value)
  };

  const data = getData();  //would probably be the data object from handleSubmit
  const transformedData = transform(data, onTransform);
```

**Reply by bluebill1049** [reply]
I think something like `Zod` is pretty much doing that now (the data transform bit).
